src/ConstructionBase.jl

```diff
@@ -43,13 +43,25 @@ end
 
 getproperties(o::NamedTuple) = o
 getproperties(o::Tuple) = o
-@generated function getproperties(obj)
-    fnames = fieldnames(obj)
-    fvals = map(fnames) do fname
-        Expr(:call, :getproperty, :obj, QuoteNode(fname))
+if VERSION >= v"1.7"
+    @inline function getproperties(obj::T) where {T}
+        fnames = propertynames(obj)
+        NamedTuple{fnames}(getproperty.(Ref(obj), fnames))
+    end
+else
+    @generated function getproperties(obj)
+        fnames = fieldnames(obj)
+        fvals = map(fnames) do fname
+            Expr(:call, :getproperty, :obj, QuoteNode(fname))
+        end
+        fvals = Expr(:tuple, fvals...)
+        quote
+            if $fnames != propertynames(obj)
```

Comment:
I think `propertynames` is really slow in old julia versions. Like `300ns`. I also think `setproperties` depends on `getproperties`. So thinks like `Setfield.@set` will get unacceptably slow.